feat(models): Triton Attention backend

cathalobrien opened the pull request:
## Description
triton backend for the transformer processor. As fast as flash-attention 2 without all the hassle to install.

I based it on the official (MIT licensed) triton fused attention [demo](https://triton-lang.org/main/getting-started/tutorials/06-fused-attention.html) and added support for sliding window. I also changed the BW pass structure to make it simpler and easier to support other attention modifications

performance tests and loss comparisons against longer runs are shown in comments below

There is a pytest suite which tests numerous different configurations compared to a reference implementation:
```
========== 144 passed, 240 skipped, 1 warning in 19.01s ==========
```

This PR changes the default attention implementation when using the transformer processor to the 'triton' backend. Since the triton backend has the same performance as flash attention v2 and does not have to be installed, this will allow users to train transformer models out-of-the-box.

I also added two env vars to allow users at inference to select the attention implementation. This is a quality of life feature a few users have suggested.
```
2026-01-13 15:33:22 INFO 'ANEMOI_INFERENCE_GRAPHTRANSFORMER_ATTENTION_BACKEND' environment variable has been set. Overwriting attention backend from triton to pyg

2026-01-13 15:27:07 INFO 'ANEMOI_INFERENCE_TRANSFORMER_ATTENTION_BACKEND' environment variable has been set. Overwriting attention backend from triton to flash_attention
```
In the future it would be better to replace the use of env vars with passing the information through the anemoi-inference config

<!-- readthedocs-preview anemoi-training start -->
----
📚 Documentation preview 📚: https://anemoi-training--716.org.readthedocs.build/en/716/

<!-- readthedocs-preview anemoi-training end -->

<!-- readthedocs-preview anemoi-graphs start -->
----
📚 Documentation preview 📚: https://anemoi-graphs--716.org.readthedocs.build/en/716/

<!-- readthedocs-preview anemoi-graphs end -->

<!-- readthedocs-preview anemoi-models start -->
----
📚 Documentation preview 📚: https://anemoi-models--716.org.readthedocs.build/en/716/

<!-- readthedocs-preview anemoi-models end -->

cathalobrien committed:
add triton attention

cathalobrien committed:
add flex attn to test correctness

cathalobrien committed:
triton benchmark harness against flash attn. results:
fused-attention-batch1-head8-d64-fwd-window=0:
     N_CTX  Triton [FP16]       Flash-2
0   1024.0      53.951905     30.154757
1   2048.0     251.862878    225.601753
2   4096.0    1003.585900    883.967242
3   8192.0    3087.151752   2919.118612
4  16384.0   14424.333821  12164.023961
fused-attention-batch1-head8-d64-fwd-window=256:
     N_CTX  Triton [FP16]      Flash-2
0   1024.0      66.118373    49.047213
1   2048.0     212.385098   224.781601
2   4096.0     627.400725   754.676149
3   8192.0    2403.518850  2072.301033
4  16384.0    5407.098626  4389.974473
fused-attention-batch1-head8-d64-bwd-window=0:
     N_CTX  Triton [FP16]      Flash-2
0   1024.0      29.038600    28.145139
1   2048.0     103.598599   107.078404
2   4096.0     221.267879   466.277452
3   8192.0     238.154238  1926.255278
4  16384.0     244.473300  7453.972862
fused-attention-batch1-head8-d64-bwd-window=256:
     N_CTX  Triton [FP16]      Flash-2
0   1024.0      31.743517    30.864637
1   2048.0      90.926897   119.329910
2   4096.0     219.786079   486.217597
3   8192.0     236.157174  1804.149605
4  16384.0     243.949151  4204.442805
fused-attention-batch1-head8-d128-fwd-window=0:
     N_CTX  Triton [FP16]       Flash-2
0   1024.0     113.015749     97.006563
1   2048.0     355.368512    378.363736
2   4096.0    1539.535832   1366.327294
3   8192.0    7007.632047   6028.029964
4  16384.0   19762.887419  17225.973294
fused-attention-batch1-head8-d128-fwd-window=256:
     N_CTX  Triton [FP16]      Flash-2
0   1024.0     103.508837    89.663668
1   2048.0     525.090419   464.968188
2   4096.0    1235.160998  1332.826567
3   8192.0    3203.617410  2936.990470
4  16384.0    6880.275897  6123.385195
fused-attention-batch1-head8-d128-bwd-window=0:
     N_CTX  Triton [FP16]      Flash-2
0   1024.0      31.435884    62.969277
1   2048.0      58.816000   240.568425
2   4096.0      81.556234   987.361971
3   8192.0      86.844913  3692.221169
4  16384.0      88.609923  8978.312907
fused-attention-batch1-head8-d128-bwd-window=256:
     N_CTX  Triton [FP16]      Flash-2
0   1024.0      31.168216    62.057629
1   2048.0      57.411547   240.228307
2   4096.0      73.313512   920.541072
3   8192.0      75.291760  2300.896123
4  16384.0      75.907589  4835.565817

cathalobrien committed:
add flex attn to benchmark tests

cathalobrien committed:
upgrade to use newest triton fusedAttn implementation

cathalobrien committed:
small bug fixes

cathalobrien committed:
SW mask wip

cathalobrien committed:
working bwd sliding window

cathalobrien committed:
small changes

cathalobrien committed:
pre-commit and license

cathalobrien committed:
move tests to its own file

cathalobrien committed:
rename file to prvent pytest error when running layers/test_attention.py

cathalobrien committed:
mark as GPU only

cathalobrien committed:
skip triton tests not on GPU

cathalobrien committed:
remove unneeded device check which crashes on CPU

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cathalobrien committed:
tests skip on CPU

cathalobrien committed:
Merge branch 'feature/triton-flash-attn' of github.com:ecmwf/anemoi-core into feature/triton-flash-attn

cathalobrien committed:
Merge branch 'main' into feature/triton-flash-attn

cathalobrien committed:
fix error in sliding window where a tiny number of elements (0.0%) would be off if window-size was not a factor of BLOCK_N

cathalobrien committed:
test o96 config against flash attn if its available

ssmmnn11 committed:
Merge branch 'main' into feature/triton-flash-attn

cathalobrien committed:
adding comments (WIP)

HCookie commented on models/src/anemoi/models/triton/attention.py:
I do not understand `triton` so this will need a review from someone who does

cathalobrien committed:
Merge branch 'main' into feature/triton-flash-attn

cathalobrien committed:
cleaning up and comments wip

cathalobrien committed:
moved

cathalobrien commented on models/tests/triton/test_triton_attention.py:
```diff
@@ -0,0 +1,152 @@
+
+# These tests for a triton fused attention algorithm are adapted from
+# the fused attention example from the Triton-lang github repo (MIT license) (Credits: OpenAI kernel team)
+# The tests have been extended to support sliding window and to test real-world problem sizes aganist flash attention if its available
```
good catch, thanks

cathalobrien committed:
comments and tidying up

cathalobrien committed:
optimise bwd pass by removing uneeded loops

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cathalobrien committed:
pre-commit

cathalobrien committed:
Merge branch 'feature/triton-flash-attn' of github.com:ecmwf/anemoi-core into feature/triton-flash-attn

cathalobrien committed:
change default attention implementation to triton and update docs

cathalobrien committed:
typo

cathalobrien committed:
greatly improve accuracy by masking before applying exponent. tests now pass against flash attn to 1e-3 in vast majority of cases

cathalobrien committed:
fix not_implemented function to work with 0.0 softcap value

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cathalobrien committed:
precommit

cathalobrien committed:
Merge branch 'feature/triton-flash-attn' of github.com:ecmwf/anemoi-core into feature/triton-flash-attn

cathalobrien committed:
Merge branch 'main' into feature/triton-flash-attn

cathalobrien committed:
added env vars to change transformer and graphtransformer backends during inference runtime

cathalobrien committed:
comments

japols commented on models/src/anemoi/models/layers/block.py:
```diff
@@ -45,7 +45,9 @@
 # Number of chunks used in inference (https://github.com/ecmwf/anemoi-core/pull/66)
 NUM_CHUNKS_INFERENCE = int(os.environ.get("ANEMOI_INFERENCE_NUM_CHUNKS", "1"))
-NUM_CHUNKS_INFERENCE_PROCESSOR = int(os.environ.get("ANEMOI_INFERENCE_NUM_CHUNKS_PROCESSOR", NUM_CHUNKS_INFERENCE))
+NUM_CHUNKS_INFERENCE_PROCESSOR = int(os.environ.get("ANEMAOI_INFERENCE_NUM_CHUNKS_PROCESSOR", NUM_CHUNKS_INFERENCE))
```
```suggestion
NUM_CHUNKS_INFERENCE_PROCESSOR = int(os.environ.get("ANEMOI_INFERENCE_NUM_CHUNKS_PROCESSOR", NUM_CHUNKS_INFERENCE))
```

cathalobrien replied:
it was the pre-commit i swear :D

cathalobrien committed:
wip adding autotuning to bwd pass to improve performance on larger head dims. complex bc more block sizes and dependacnies between them due to sharing a gird. wont run but pytest which doesnt autotune passes

japols reviewed (changes requested):
Nice work!

Some comments on dtypes and a few minor changes.

cathalobrien committed:
Update models/src/anemoi/models/triton/attention.py

Co-authored-by: Jan Polster <jan.polster@ecmwf.int>

cathalobrien committed:
Update models/src/anemoi/models/layers/block.py

Co-authored-by: Jan Polster <jan.polster@ecmwf.int>

cathalobrien committed:
autotuning bwd working btu not ideal

cathalobrien committed:
pre-commit

cathalobrien committed:
Merge branch 'feature/triton-flash-attn' of github.com:ecmwf/anemoi-core into feature/triton-flash-attn

cathalobrien committed:
support bf16. correctness tests added. benchmarks show same performance as fp16

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cathalobrien committed:
raise not implemented error for causal

cathalobrien committed:
Merge branch 'feature/triton-flash-attn' of github.com:ecmwf/anemoi-core into feature/triton-flash-attn

cathalobrien committed:
split bwd into two seperate kernels to allow each to choose optimal grid via autotuning

cathalobrien committed:
Added tensor descriptors and warp spec to bwd_dkdv, reducing memory usage

cathalobrien committed:
added tensor descriptors to bwd_dq

cathalobrien committed:
rename fwd blocks to be consistent with bwd

cathalobrien committed:
cache autotuning results to dsik...allegedly

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cathalobrien committed:
check attention backend env vars during runtime

cathalobrien committed:
Merge branch 'feature/triton-flash-attn' of github.com:ecmwf/anemoi-core into feature/triton-flash-attn

cathalobrien committed:
rename variables for clarity

cathalobrien committed:
check input tensors are contiguous

ssmmnn11 reviewed (changes requested):
nice!

cathalobrien commented on models/src/anemoi/models/triton/attention.py:
```diff
@@ -0,0 +1,908 @@
+        RCP_LN2 = 1.4426950408889634  # = 1.0 / ln(2)
+        k = k * (ctx.sm_scale * RCP_LN2)
+        PRE_BLOCK = 128
+        assert N_CTX % PRE_BLOCK == 0
```
yeah we should check blocks in fwd divide evenly into N_CTX or mask them

cathalobrien committed:
wip addressing comments and investigating bug

cathalobrien committed:
tests passed by lowering precision requirement for bf16

cathalobrien committed:
@HCookie comments

HCookie committed:
Merge branch 'main' into feature/triton-flash-attn

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cathalobrien committed:
fix bug in sliding window with compiler bounds

cathalobrien committed:
uneven ctx working

cathalobrien committed:
uneven CTX works in the fwd! but its 50% slower then flash attn when working on uneven ctx and the kernel uses more memory

cathalobrien committed:
uncomment padding code

cathalobrien committed:
bwd uneven ctx wip

cathalobrien committed:
bwd wip

cathalobrien committed:
bwd global seems to work for uneven ctx

cathalobrien committed:
works for uneven ctx

cathalobrien committed:
Merge branch 'feature/triton-flash-attn' of github.com:ecmwf/anemoi-core into feature/triton-flash-attn

cathalobrien committed:
increase ctx padding to max block size to support all block sizes offered by autotune

cathalobrien committed:
remove uneeded pruning and static asserts since we support all ctx lengths now

cathalobrien committed:
pre-commit

cathalobrien commented:
update: FINALLY it works with uneven contexts. 

Unfortunately there's an overhead of dealing with uneven contexts, but now that it works hopefully the performance can be improved.

cathalobrien commented on models/src/anemoi/models/schemas/decoder.py:
```diff
@@ -60,8 +60,8 @@ class TransformerDecoderSchema(TransformerModelComponent):
     "Dropout probability used for multi-head self attention, default 0.0"
-    attention_implementation: str = Field(example="flash_attention")
-    "Attention implementation to use. Default to 'flash_attention'."
+    attention_implementation: str = Field(example="triton_attention")
```
ah ok, thanks I have updated the comment to reflect that

cathalobrien committed:
pre-commit

cathalobrien committed:
flash-attn like tests and fix bug in causal masking sign

cathalobrien committed:
masking uneven ctx measures gives us back speedup in even ctx case. still too slow in uneven case
=== Benchmarking: o96-512c fwd ===
Z=1 H=16 N_CTX=40333 HEAD_DIM=32

Flash-Attention 2:
flash_attn_func
        elapsed time: 1.23 ms / iter
        peak memory usage: 829.46 MB

 Triton Attention:
apply
        elapsed time: 1.78 ms / iter
        peak memory usage: 1049.97 MB

=== Benchmarking: o96-1024c fwd ===
Z=1 H=16 N_CTX=40333 HEAD_DIM=64

Flash-Attention 2:
flash_attn_func
        elapsed time: 2.01 ms / iter
        peak memory usage: 910.80 MB

 Triton Attention:
apply
        elapsed time: 2.90 ms / iter
        peak memory usage: 1462.57 MB

cathalobrien committed:
speedup in even ctx bwd pass case by adding more compile time constant guards

=== Benchmarking: o96-512c fwd+bwd ===
Z=1 H=16 N_CTX=40320 HEAD_DIM=32

Flash-Attention 2:
flash_attn_func
        elapsed time: 4.08 ms / iter
        peak memory usage: 1034.42 MB

 Triton Attention:
apply
        elapsed time: 3.40 ms / iter
        peak memory usage: 1253.51 MB

=== Benchmarking: o96-1024c fwd+bwd ===
Z=1 H=16 N_CTX=40320 HEAD_DIM=64

Flash-Attention 2:
flash_attn_func
        elapsed time: 6.41 ms / iter
        peak memory usage: 1541.67 MB

 Triton Attention:
apply
        elapsed time: 4.62 ms / iter
        peak memory usage: 1720.88 MB

=== Benchmarking: o96-512c fwd+bwd ===
Z=1 H=16 N_CTX=40333 HEAD_DIM=32

Flash-Attention 2:
flash_attn_func
        elapsed time: 4.15 ms / iter
        peak memory usage: 1035.71 MB

 Triton Attention:
apply
        elapsed time: 6.50 ms / iter
        peak memory usage: 1494.91 MB

=== Benchmarking: o96-1024c fwd+bwd ===
Z=1 H=16 N_CTX=40333 HEAD_DIM=64

Flash-Attention 2:
flash_attn_func
        elapsed time: 6.43 ms / iter
        peak memory usage: 1542.36 MB

 Triton Attention:
apply
        elapsed time: 9.81 ms / iter
        peak memory usage: 2202.28 MB

cathalobrien committed:
reduce padding in fwd uneven ctx case

cathalobrien committed:
spedup uneven ctx fwd pass by removing most padding

=== Benchmarking: o96-512c fwd ===
Z=1 H=16 N_CTX=40333 HEAD_DIM=32

Flash-Attention 2:
flash_attn_func
        elapsed time: 1.24 ms / iter
        peak memory usage: 829.46 MB

 Triton Attention:
apply
        elapsed time: 1.51 ms / iter
        peak memory usage: 930.47 MB

=== Benchmarking: o96-1024c fwd ===
Z=1 H=16 N_CTX=40333 HEAD_DIM=64

Flash-Attention 2:
flash_attn_func
        elapsed time: 2.01 ms / iter
        peak memory usage: 910.80 MB

 Triton Attention:
apply
        elapsed time: 2.39 ms / iter
        peak memory usage: 1085.57 MB
"

cathalobrien committed:
speedup bwd uneven ctx case by removing most padding

=== Benchmarking: o96-512c fwd+bwd ===
Z=1 H=16 N_CTX=40333 HEAD_DIM=32

Flash-Attention 2:
flash_attn_func
        elapsed time: 4.14 ms / iter
        peak memory usage: 1035.71 MB

 Triton Attention:
apply
        elapsed time: 5.49 ms / iter
        peak memory usage: 1253.57 MB

=== Benchmarking: o96-1024c fwd+bwd ===
Z=1 H=16 N_CTX=40333 HEAD_DIM=64

Flash-Attention 2:
flash_attn_func
        elapsed time: 6.44 ms / iter
        peak memory usage: 1542.36 MB

 Triton Attention:
apply
        elapsed time: 7.97 ms / iter
        peak memory usage: 1721.61 MB

cathalobrien committed:
speedup uneven ctx by guarding more masks

cathalobrien committed:
add test for kernel determinism

cathalobrien committed:
pre-commit

cathalobrien committed:
add comments to forward pass and supporting functions

cathalobrien committed:
added comments to backward pass

cathalobrien commented:
Speedup is back after adding compile-time guards against extra logic

```
=== Benchmarking: o96-512c fwd+bwd ===
Z=1 H=16 N_CTX=40320 HEAD_DIM=32

Flash-Attention 2:
flash_attn_func
        elapsed time: 4.08 ms / iter
        peak memory usage: 1034.42 MB


 Triton Attention:
apply
        elapsed time: 3.14 ms / iter
        peak memory usage: 1253.51 MB



=== Benchmarking: o96-1024c fwd+bwd ===
Z=1 H=16 N_CTX=40320 HEAD_DIM=64

Flash-Attention 2:
flash_attn_func
        elapsed time: 6.41 ms / iter
        peak memory usage: 1541.67 MB


 Triton Attention:
apply
        elapsed time: 4.37 ms / iter
        peak memory usage: 1720.88 MB



=== Benchmarking: o96-512c fwd+bwd ===
Z=1 H=16 N_CTX=40333 HEAD_DIM=32

Flash-Attention 2:
flash_attn_func
        elapsed time: 4.14 ms / iter
        peak memory usage: 1035.71 MB


 Triton Attention:
apply
        elapsed time: 4.36 ms / iter
        peak memory usage: 1253.57 MB



=== Benchmarking: o96-1024c fwd+bwd ===
Z=1 H=16 N_CTX=40333 HEAD_DIM=64

Flash-Attention 2:
flash_attn_func
        elapsed time: 6.42 ms / iter
        peak memory usage: 1542.36 MB


 Triton Attention:
apply
        elapsed time: 6.05 ms / iter
        peak memory usage: 1721.61 MB
```

cathalobrien committed:
small speedup and memory savings on larger problem sizes by moving bwd k scaling into kernel

cathalobrien committed:
Merge branch 'main' into feature/triton-flash-attn

cathalobrien committed:
fix bug with device side tensor descriptors

cathalobrien committed:
varlen fwd ok

cathalobrien committed:
make inputs contiguous if needed

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

icedoom888 committed:
Merge branch 'main' into feature/triton-flash-attn

icedoom888 committed:
Merge branch 'feature/triton-flash-attn' of github.com:ecmwf/anemoi-core into feature/triton-flash-attn

icedoom888 committed:
Merge branch 'main' into feature/triton-flash-attn

radiradev commented:
Hi @cathalobrien, flash-attention 4 has been released, perhaps your benchmarks should be updated to compare against that? 
<img width="1243" height="620" alt="image" src="https://github.com/user-attachments/assets/e0808988-a60d-49dd-aae2-1f362551b6c1" />
Also, it has direct support for flex attention.

ssmmnn11 committed:
Merge branch 'main' into feature/triton-flash-attn

cathalobrien committed:
varlen bwd wip

cathalobrien committed:
sliding window wip

cathalobrien committed:
remove varlen until attention is merged

cathalobrien committed:
merge

cathalobrien committed:
comments from ai

cathalobrien committed:
typo

cathalobrien committed:
cumulative loss test

cathalobrien committed:
merge main

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

HCookie committed:
Merge branch 'main' into feature/triton-flash-attn

HCookie commented:
@cathalobrien plz I need this code

anaprietonem commented:
While this is a very useful feature, @cathalobrien noticed that more work is required as at the moment we are seeing stability issues that mean the model doesn't converge in the same way as when flash-attention is used.
